### ISSUE TITLE
Sidecar is not injected as first container when `holdApplicationUntilProxyStarts` is set

### ISSUE THREAD

**Issue by thib92** [istio/istio#33911]
**Bug description**

Thanks to #24737, we can set `holdApplicationUntilProxyStarts` to inject a `postStart` hook in the sidecar container to wait for the Envoy proxy to be ready, and then unblock the actual entrypoints of the other containers to run, fixing issues where a race condition would have the main container start trying to connect to external services before the Envoy sidecar proxy is ready (creating "Connection Refused" errors).

This works because `kubelet` calls each `postStart` hook synchronously, one by one, in the order of the containers in the list of the `Pod` spec. **This means that the sidecar should be the first container of the pod.**

#24737 implemented [`injectContainers`](https://github.com/istio/istio/pull/24737/files#diff-3658c8df7815d0014a39c664de6ee52d83c4e85d6f57d2eadfc829c008e87bf1R814-R836) that doesn't just append the sidecar container, but actually inserts it in first instead of last.

However, looking at the main branch of the repo, this function does not exist anymore.

Looking at [the history for the file](https://github.com/istio/istio/commits/master?before=c4af985cbef0639403a5ceb0bb4cdfe64f0eabb1+35&branch=master&path%5B%5D=pkg&path%5B%5D=kube&path%5B%5D=inject&path%5B%5D=inject.go), it seems like #26596 introduced the regression.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

The sidecar should be injected in first in the container list, under the condition that `holdApplicationUntilProxyStarts` is set. Otherwise there are unexpected behaviors explained in #24737 (`kubectl exec` would default to the first container, aka the proxy).

**Steps to reproduce the bug**

* Create a pod with sidecar injection enabled
* Notice that the sidecar is not in first place in the containers list

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

Kubernetes:
```
Client Version: v1.19.7
Server Version: v1.16.15-eks-cf75c9
```

Istio: 1.9.1

**How was Istio installed?**

Helm charts

**Environment where the bug was observed (cloud vendor, OS, etc)**

AWS EKS

**Comment by howardjohn** [comment]
using the config:
```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello
      tier: backend
      track: stable
  template:
    metadata:
      labels:
        app: hello
        tier: backend
        track: stable
      annotations:
        proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": false }'
    spec:
      containers:
      - name: hello
        image: "fake.docker.io/google-samples/hello-go-gke:1.0"
        ports:
        - name: http
          containerPort: 80
        livenessProbe:
          httpGet:
            port: http
        readinessProbe:
          httpGet:
            port: 3333
```

I see it injected first. Can you describe more how you reproduced this?

**Comment by thib92** [comment]
Interesting. I am installing Istio through Helm. Here is my `values.yaml` file for `istiod`:

```yaml
meshConfig:
  defaultConfig:
    holdApplicationUntilProxyStarts: true
```

After applying the update, I annotated my namespace to enable sidecar injection, and deleted one by one my pods to have the injection on new pods, and it was injected last.

Let me try with a `proxy.istio.io/config` annotation and see if it works.

Did you try with `1.9.1`?

**Comment by Laci21** [comment]
Hi @thib92 and @howardjohn,

@thib92, do you happen to have multiple containers in your deployment spec where you see this issue?

The reason I'm asking is that we've just experienced the following issue starting from Istio 1.9 (does not happen with Istio 1.8):

With **Istio 1.9** when the **deployment spec has multiple containers** and there are **annotations set under `spec.template.metadata.annotations` after the sidecar injection** happens and in the pod spec we have the `istio-init` init container and the `istio-proxy` container the **annotations are not present on the pod itself**.

This could cause a number of totally different issues based on what those annotations are supposed to be used for. It could cause this `holdApplicationUntilProxyStarts` issue as well, for us when we first encountered this a `vault-injection` annotation was missing and hence the vault container injection did not happen.

**Comment by thib92** [comment]
Hey @Laci21 , we don't have any additional containers for the pods for which I've seen the issue, only one container.
Moreover, I saw here that using the Pod annotation actually works, but the `defaultConfig` doesn't. I'm still investigating why the setting wasn't picked up.

**Comment by thib92** [comment]
Okay I figured it out.
Turns out an update to the `istio-discovery` Helm chart which only updates values in the `ConfigMap` will not trigger a rolling update of `istiod` (makes sense), so it did not pick up the new configuration.

I triggered a rolling restart of `istiod` with `kubectl rollout restart istiod -n istio-system`, and then killed my pod, and the newly created pod did get it first.

**Comment by thib92** [comment]
I restarted my `istiod` pod, thinking that it didn't grab the configuration change, but the issue is still present.
This is what my `istio` `ConfigMap` looks like:

```yaml
apiVersion: v1
data:
  mesh: |-
    defaultConfig:
      discoveryAddress: istiod.istio-system.svc:15012
      holdApplicationUntilProxyStarts: true
      tracing:
        zipkin:
          address: zipkin.istio-system:9411
    enablePrometheusMerge: true
    rootNamespace: null
    trustDomain: cluster.local
  meshNetworks: 'networks: {}'
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: istiod
    meta.helm.sh/release-namespace: istio-system
  creationTimestamp: "2021-03-30T13:16:57Z"
  labels:
    app.kubernetes.io/managed-by: Helm
    install.operator.istio.io/owning-resource: unknown
    istio.io/rev: default
    operator.istio.io/component: Pilot
    release: istiod
  name: istio
  namespace: istio-system
  resourceVersion: "193305169"
  selfLink: /api/v1/namespaces/istio-system/configmaps/istio
  uid: 7a7e8132-b75a-47d4-9bcc-63b8cefbfe2a

```

**Comment by thib92** [comment]
Upgrading to Istio 1.10.2 fixed the issue.
I'm not sure if it's because the bug was fixed between 1.9.1 and 1.10.2, or if the in-place upgrade of Istio fixed it, but I can now see the sidecar being injected first.
